src/Castle.Facilities.AspNetCore/WindsorRegistrationExtensions.cs

```diff
@@ -0,0 +1,213 @@
+		/// </summary>
+		/// <param name="services">ASP.NET Core service collection from Microsoft.Extensions.DependencyInjection</param>
+		/// <param name="container">Windsor container which activators call resolve against</param>
+		/// <param name="configure">Configuration options for controlling registration and lifestyles of controllers, tagHelpers and viewComponents</param>
+		/// <param name="serviceProviderFactory">Optional factory for creating a custom service provider</param>
+		public static IServiceProvider AddWindsor(this IServiceCollection services, IWindsorContainer container, Action<WindsorRegistrationOptions> configure = null, Func<IServiceProvider> serviceProviderFactory = null)
+		{
+			AddFrameworkResolversToWindsor(services, container);
+			InstallFrameworkIntegration(services, container);
+
+			var options = new WindsorRegistrationOptions();
+			configure?.Invoke(options);
+
+			AddApplicationComponentsToWindsor(container, options);
+
+			var serviceProvider = serviceProviderFactory?.Invoke() ?? services.BuildServiceProvider(validateScopes: false);
+			SetServiceProviderForFrameworkResolvers(serviceProvider, container);
+			return serviceProvider;
+		}
+
+		/// <summary>
+		/// For making types available to the IServiceCollection using 'late bound' factories which resolve from Windsor. This makes things like the @Inject directive in Razor work.
+		/// </summary>
+		/// <param name="registration">The component registration that gets copied across to the IServiceCollection</param>
+		/// <param name="serviceType">The type of the service, we could use T but sometimes the compiler infers this as type 'object'</param>
+		/// <param name="container">Windsor container which 'activators'/'cross wired' components resolve from</param>
+		/// <param name="serviceCollection">ASP.NET Core service collection from Microsoft.Extensions.DependencyInjection</param>
+		/// <param name="lifestyleType">The lifestyle type that needs to be copied across, only Singleton, Scoped and Transient are supported</param>
+		public static ComponentRegistration<T> CrossWired<T>(this ComponentRegistration<T> registration, Type serviceType, IWindsorContainer container, IServiceCollection serviceCollection, LifestyleType lifestyleType = LifestyleType.Scoped) where T : class
+		{
+			registration.LifeStyle.Is(lifestyleType);
+
+			if (lifestyleType == LifestyleType.Transient)
+			{
+				serviceCollection.AddTransient(serviceType, p =>
+				{
+					return container.Resolve(serviceType);
+				});
+			}
+			else if (lifestyleType == LifestyleType.Scoped)
+			{
+				registration.Activator<CrossWiredComponentActivator>();
+				serviceCollection.AddScoped(serviceType, p =>
+				{
+					return container.Resolve(serviceType);
+				});
+			}
+			else if (lifestyleType == LifestyleType.Singleton)
+			{
+				registration.Activator<CrossWiredComponentActivator>();
```

Comment:
Singleton cross wired components are released by disposing the ServiceProvider not the WindsorContainer.